ocaml/xapi/repository_helpers.ml

```diff
@@ -459,22 +459,70 @@ let parse_updateinfo_list acc line =
       debug "Ignore unrecognized line '%s' in parsing updateinfo list" line ;
       acc
 
-let get_updates_from_updateinfo ~__context repositories =
-  let params_of_updateinfo_list =
+let is_obsoleted pkg_name enabled_repos =
+  let params =
+    [
+      "-a"
+    ; "--plugins"
+    ; "--disablerepo=*"
+    ; Printf.sprintf "--enablerepo=%s" enabled_repos
+    ; "--whatobsoletes"
+    ; pkg_name
+    ; "--qf"
+    ; "%{name}"
+    ]
+  in
+  match
+    Helpers.call_script !Xapi_globs.repoquery_cmd params
+    |> Astring.String.cuts ~sep:"\n"
+    |> List.filter (fun x -> not (Astring.String.is_empty x))
+  with
+  | [] ->
+      false
+  | _ ->
+      debug "Available package %s has been obsoleted." pkg_name ;
+      true
+  | exception _ ->
+      warn
+        "Can't determine if available package %s is obsoleted or not. Assuming \
+         it is not."
+        pkg_name ;
+      false
+
+let run_yum_updateinfo_list sub_command enabled_repos =
+  let params =
     [
       "-q"
     ; "--disablerepo=*"
-    ; Printf.sprintf "--enablerepo=%s" (String.concat "," repositories)
+    ; Printf.sprintf "--enablerepo=%s" enabled_repos
     ; "updateinfo"
     ; "list"
-    ; "updates"
+    ; sub_command
     ]
   in
-  Helpers.call_script !Xapi_globs.yum_cmd params_of_updateinfo_list
+  Helpers.call_script !Xapi_globs.yum_cmd params
   |> assert_yum_error
   |> Astring.String.cuts ~sep:"\n"
+  |> List.map (fun x ->
+         debug "yum updateinfo list %s: %s" sub_command x ;
+         x
+     )
   |> List.fold_left parse_updateinfo_list []
 
+let get_updates_from_updateinfo ~__context repositories =
```

Comment:
Would this be better expressed with strings sets? `module StringSet = Set.Make(String)`

Reply:
The `repositories` comes from query on XAPI DB. So I think it should keep  the uniqueness of elements well. And usually there are only two repositories in this list.